File: packages/figma-downloader-ts/src/chunk.mjs
// chunk.mjs — plan and execute the chunked download.
//
// The skeleton (a depth-truncated `/files/:key`) carries the document tree down
// to the truncation frontier. The frontier nodes — those at exactly `depth` —
// are the roots we re-fetch in full via `/nodes`, a batch of ids per request,
// each response streamed to its own part file on disk. Reconstruction grafts
// those subtrees back onto the skeleton.
//
// Two scaling mechanisms beyond plain frontier chunking:
//   • Content-hash part files (TR-12): a part is named by the sha256 of its
//     sorted id-batch, so an unchanged batch yields the same filename and a
//     resume reuses it. The manifest records, per planned group, which part
//     files it produced — so resume is correct even when a group was re-chunked.
//   • Byte/time-aware re-chunking (TR-3): if a fetched batch exceeds its byte or
//     time budget, its oversized subtrees are split in memory into a SHALLOW
//     parent part (one-level children) plus a full part per child, recursively.
//     No single part file then exceeds the budget — the guarantee that makes the
//     downloader work for a file of ANY size, regardless of the initial tier.

import { promises as fs, createWriteStream } from 'node:fs';
import path from 'node:path';
import { fetchNodes } from './figmaClient.mjs';
import { batch, pool, sha256OfBatch, stringify, byteLength } from './util.mjs';

/**
 * Walk the skeleton document and collect the ids of every node at exactly
 * `depth` (document = 0, pages/CANVAS = 1, top frames = 2). These are the chunk
 * roots.
 *
 * @param {object} skeleton  a `/files/:key?depth=N` response
 * @param {number} depth
 * @returns {string[]} frontier node ids, document order, de-duplicated
 */
export function collectFrontierIds(skeleton, depth) {
  const doc = skeleton?.document;
  if (!doc) throw new Error('skeleton has no `document` — cannot plan chunks.');
  const ids = [];
  (function walk(node, level) {
    if (!node || typeof node !== 'object') return;
    if (level === depth) {
      if (node.id) ids.push(node.id);
      return; // do not descend past the frontier
    }
    for (const child of node.children || []) walk(child, level + 1);
  })(doc, 0);
  return [...new Set(ids)];
}

/** Drop a node's `children` array, keeping everything else (a graft stub). */
function stripChildren(node) {
  const { children, ...rest } = node;
  return rest;
}

async function exists(p) {
  try {
    await fs.access(p);
    return true;
  } catch {
    return false;
  }
}

/** Write a JSON string to disk — buffered, or streamed with backpressure. */
async function writeJson(partPath, json, stream) {
  if (!stream) {
    await fs.writeFile(partPath, json, 'utf8');
    return;
  }
  await new Promise((resolve, reject) => {
    const ws = createWriteStream(partPath, { encoding: 'utf8' });
    ws.on('error', reject);
    ws.on('finish', resolve);
    const CH = 1 << 16;
    let i = 0;
    (function pump() {
      let ok = true;
      while (i < json.length && ok) {
        const slice = json.slice(i, i + CH);
        i += CH;
        if (i >= json.length) {
          ws.end(slice);
          return;
        }
        ok = ws.write(slice);
      }
      if (i < json.length && !ok) ws.once('drain', pump);
    })();
  });
}

/**
 * Stream-download the full subtree for every frontier id into content-hashed
 * part files under `<workDir>/parts/`, re-chunking any over-budget subtree, and
 * write a complete `manifest.json` in a single pass.
 *
 * @returns {Promise<{ manifest: object, partFiles: string[] }>}
 */
export async function downloadChunks({
  fileKey,
  token,
  frontierIds,
  depth,
  workDir,
  size,
  batchSize = 10,
  concurrency = 4,
  resume = false,
  retries = 3,
  timeoutMs = 60_000,
  geometry = false,
  version,
  nodeDepth,
  stream = false,
  rechunk = false,
  maxPartBytes = 0,
  maxPartMs = 0,
  maxRechunkDepth = 4,
  limiter = null,
  log = () => {},
}) {
  const partsDir = path.join(workDir, 'parts');
  await fs.mkdir(partsDir, { recursive: true });

  const batches = batch(frontierIds, batchSize);

  // Prior manifest enables correct resume even for re-chunked groups.
  const priorGroups = new Map();
  if (resume) {
    try {
      const prior = JSON.parse(await fs.readFile(path.join(workDir, 'manifest.json'), 'utf8'));
      for (const g of prior.groups || []) priorGroups.set(g.sha, g);
    } catch {
      /* no prior manifest — nothing to resume */
    }
  }

  const partsMeta = [];
  const groupsMeta = new Array(batches.length);
  const writtenFiles = new Set();
  let written = 0;
  let skipped = 0;
  let rechunked = 0;
  const oversizedLeaves = [];

  const mapsOf = (entry) => ({
    components: entry?.components || {},
    componentSets: entry?.componentSets || {},
    styles: entry?.styles || {},
  });
  const nodePart = (node, maps) => ({ nodes: { [node.id]: { document: node, ...maps } } });

  const writePart = async (file, payload, ids, elapsedMs = 0) => {
    const partPath = path.join(partsDir, file);
    const json = stringify(payload);
    const bytes = byteLength(json);
    await writeJson(partPath, json, stream);
    if (!writtenFiles.has(file)) {
      writtenFiles.add(file);
      partsMeta.push({ file, ids: [...ids], sha: file.slice(5, -5), bytes, elapsedMs });
      written += 1;
    }
    return bytes;
  };

  // Split one over-budget node into a shallow parent part + full child parts.
  const splitNode = async (node, level, maps, groupFiles) => {
    const bytes = byteLength(stringify(node));
    const kids = Array.isArray(node.children) ? node.children : [];
    const over = maxPartBytes > 0 && bytes > maxPartBytes;
    const canSplit = rechunk && level < maxRechunkDepth && kids.length > 0;
    if (!over || !canSplit) {
      if (over && kids.length === 0 && !oversizedLeaves.includes(node.id)) oversizedLeaves.push(node.id);
      const file = `part-${sha256OfBatch([node.id])}.json`;
      await writePart(file, nodePart(node, maps), [node.id]);
      groupFiles.push(file);
      return;
    }
    rechunked += 1;
    const shallow = { ...node, children: kids.map(stripChildren) };
    const file = `part-${sha256OfBatch([node.id])}.json`;
    await writePart(file, nodePart(shallow, maps), [node.id]);
    groupFiles.push(file);
    for (const child of kids) await splitNode(child, level + 1, maps, groupFiles);
  };

  const tasks = batches.map((ids, bi) => async () => {
    const gsha = sha256OfBatch(ids);
    const groupFiles = [];

    // --- Resume: reuse a prior group's parts if all are present. ---
    if (resume && priorGroups.has(gsha)) {
      const g = priorGroups.get(gsha);
      const parts = g.parts || [];
      const present = parts.length && (await Promise.all(parts.map((f) => exists(path.join(partsDir, f))))).every(Boolean);
      if (present) {
        skipped += 1;
        for (const f of parts) {
          if (!writtenFiles.has(f)) {
            writtenFiles.add(f);
            const st = await fs.stat(path.join(partsDir, f));
            partsMeta.push({ file: f, ids: g.ids || ids, sha: f.slice(5, -5), bytes: st.size, elapsedMs: 0 });
          }
          groupFiles.push(f);
        }
        groupsMeta[bi] = { sha: gsha, ids, parts: groupFiles, rechunked: !!g.rechunked, resumed: true };
        log(`  [${bi + 1}/${batches.length}] resume: skip group ${gsha} (${groupFiles.length} part(s))`);
        return;
      }
    }

    // --- Fetch the batch. ---
    const t0 = Date.now();
    const data = await fetchNodes(fileKey, ids, token, {
      retries,
      timeoutMs,
      geometry,
      version,
      nodeDepth,
      onRetry: ({ status, attempt, waitMs }) =>
        log(`  [${bi + 1}/${batches.length}] retry #${attempt} (status ${status}) in ${waitMs}ms`),
    });
    const elapsedMs = Date.now() - t0;
    const batchBytes = byteLength(stringify(data));
    const overBudget =
      rechunk && ((maxPartBytes > 0 && batchBytes > maxPartBytes) || (maxPartMs > 0 && elapsedMs > maxPartMs));

    if (!overBudget) {
      const file = `part-${gsha}.json`;
      await writePart(file, data, ids, elapsedMs);
      groupFiles.push(file);
      groupsMeta[bi] = { sha: gsha, ids, parts: groupFiles, rechunked: false };
      const got = Object.keys(data.nodes || {}).length;
      log(`  [${bi + 1}/${batches.length}] wrote ${file} — ${got}/${ids.length} node(s), ${batchBytes}B`);
      return;
    }

    // --- Over budget: re-chunk each subtree in memory. ---
    log(`  [${bi + 1}/${batches.length}] over budget (${batchBytes}B / ${elapsedMs}ms) — re-chunking`);
    for (const id of ids) {
      const entry = data.nodes?.[id];
      if (!entry?.document) continue;
      await splitNode(entry.document, depth, mapsOf(entry), groupFiles);
    }
    groupsMeta[bi] = { sha: gsha, ids, parts: groupFiles, rechunked: true };
  });

  await pool(tasks, concurrency, { limiter });

  partsMeta.sort((a, b) => (a.file < b.file ? -1 : a.file > b.file ? 1 : 0));
  const partFiles = partsMeta.map((p) => path.join(partsDir, p.file));

  const manifest = {
    fileKey,
    size: size ?? null,
    frontierDepth: depth,
    batchSize,
    concurrency,
    idCount: frontierIds.length,
    batchCount: batches.length,
    downloaded: written,
    resumedSkipped: skipped,
    rechunked,
    oversizedLeaves,
    frontierIds: [...frontierIds],
    parts: partsMeta,
    groups: groupsMeta.filter(Boolean),
  };
  await fs.writeFile(path.join(workDir, 'manifest.json'), JSON.stringify(manifest, null, 2), 'utf8');
  return { manifest, partFiles };
}

File: packages/py/figma_downloader/chunk.py
"""chunk.py — plan and execute the chunked download.

Twin of ts/src/chunk.mjs. Content-hash part files (TR-12),
byte/time-aware re-chunking (TR-3), manifest-aware resume, and a single-pass
manifest write. The pool runs batches concurrently; each batch task returns its
local results which are aggregated in batch order (deterministic, lock-free).
"""

from __future__ import annotations

import json
import os
import time
from typing import Any, Callable

from .client import fetch_nodes
from .util import batch, byte_length, pool, sha256_of_batch, stringify


def collect_frontier_ids(skeleton: dict, depth: int) -> list[str]:
    """Collect ids of every node at exactly `depth` (the chunk roots)."""
    doc = (skeleton or {}).get("document")
    if not doc:
        raise ValueError("skeleton has no `document` — cannot plan chunks.")
    ids: list[str] = []

    def walk(node: Any, level: int) -> None:
        if not isinstance(node, dict):
            return
        if level == depth:
            if node.get("id"):
                ids.append(node["id"])
            return
        for child in node.get("children") or []:
            walk(child, level + 1)

    walk(doc, 0)
    seen = set()
    out = []
    for i in ids:
        if i not in seen:
            seen.add(i)
            out.append(i)
    return out


def _strip_children(node: dict) -> dict:
    return {k: v for k, v in node.items() if k != "children"}


def download_chunks(
    *,
    file_key: str,
    token: str,
    frontier_ids: list[str],
    depth: int,
    work_dir: str,
    size: str | None = None,
    batch_size: int = 10,
    concurrency: int = 4,
    resume: bool = False,
    retries: int = 3,
    timeout_ms: int = 60_000,
    geometry: bool = False,
    node_depth: int | None = None,
    stream: bool = False,
    rechunk: bool = False,
    max_part_bytes: int = 0,
    max_part_ms: int = 0,
    max_rechunk_depth: int = 4,
    limiter=None,
    log: Callable[..., None] = lambda *a: None,
) -> dict:
    parts_dir = os.path.join(work_dir, "parts")
    os.makedirs(parts_dir, exist_ok=True)
    batches = batch(frontier_ids, batch_size)

    prior_groups: dict[str, dict] = {}
    if resume:
        try:
            with open(os.path.join(work_dir, "manifest.json"), encoding="utf-8") as fh:
                prior = json.load(fh)
            for g in prior.get("groups", []):
                prior_groups[g["sha"]] = g
        except (OSError, ValueError, KeyError):
            pass

    def maps_of(entry: dict) -> dict:
        return {
            "components": entry.get("components") or {},
            "componentSets": entry.get("componentSets") or {},
            "styles": entry.get("styles") or {},
        }

    def node_part(node: dict, maps: dict) -> dict:
        return {"nodes": {node["id"]: {"document": node, **maps}}}

    def write_json(part_path: str, json_str: str) -> None:
        with open(part_path, "w", encoding="utf-8") as f:
            if not stream:
                f.write(json_str)
            else:
                ch = 1 << 16
                for i in range(0, len(json_str), ch):
                    f.write(json_str[i : i + ch])

    def process(args):
        ids, bi = args
        gsha = sha256_of_batch(ids)
        group_files: list[str] = []
        local_parts: list[dict] = []
        oversized: list[str] = []
        counters = {"rechunked": 0}
        written_local: set[str] = set()

        def emit(file: str, payload, ids2, elapsed=0.0, from_disk=False, disk_bytes=0):
            if file in written_local:
                return
            written_local.add(file)
            sha = file[5:-5]
            if from_disk:
                local_parts.append({"file": file, "ids": list(ids2), "sha": sha, "bytes": disk_bytes, "elapsedMs": 0, "fresh": False})
            else:
                json_str = stringify(payload)
                b = byte_length(json_str)
                write_json(os.path.join(parts_dir, file), json_str)
                local_parts.append({"file": file, "ids": list(ids2), "sha": sha, "bytes": b, "elapsedMs": elapsed, "fresh": True})

        # --- Resume: reuse a prior group's parts if all present. ---
        if resume and gsha in prior_groups:
            g = prior_groups[gsha]
            parts = g.get("parts", [])
            if parts and all(os.path.exists(os.path.join(parts_dir, f)) for f in parts):
                for f in parts:
                    st = os.stat(os.path.join(parts_dir, f))
                    emit(f, None, g.get("ids", ids), from_disk=True, disk_bytes=st.st_size)
                    group_files.append(f)
                log(f"  [{bi + 1}/{len(batches)}] resume: skip group {gsha} ({len(group_files)} part(s))")
                return {"sha": gsha, "ids": ids, "parts": group_files, "rechunked": bool(g.get("rechunked")),
                        "resumed": True, "localParts": local_parts, "localRechunked": 0, "oversized": []}

        # --- Fetch the batch. ---
        if limiter:
            limiter.acquire()
        t0 = time.monotonic()
        data = fetch_nodes(
            file_key, ids, token,
            retries=retries, timeout_ms=timeout_ms, geometry=geometry, node_depth=node_depth,
            on_retry=lambda info: log(f"  [{bi + 1}/{len(batches)}] retry #{info['attempt']} (status {info['status']}) in {info['waitMs']}ms"),
        )
        elapsed = (time.monotonic() - t0) * 1000
        batch_bytes = byte_length(stringify(data))
        over = rechunk and ((max_part_bytes > 0 and batch_bytes > max_part_bytes) or (max_part_ms > 0 and elapsed > max_part_ms))

        if not over:
            file = f"part-{gsha}.json"
            emit(file, data, ids, elapsed)
            group_files.append(file)
            got = len((data.get("nodes") or {}))
            log(f"  [{bi + 1}/{len(batches)}] wrote {file} — {got}/{len(ids)} node(s), {batch_bytes}B")
            return {"sha": gsha, "ids": ids, "parts": group_files, "rechunked": False,
                    "resumed": False, "localParts": local_parts, "localRechunked": 0, "oversized": []}

        # --- Over budget: re-chunk each subtree in memory. ---
        log(f"  [{bi + 1}/{len(batches)}] over budget ({batch_bytes}B / {elapsed:.0f}ms) — re-chunking")

        def split_node(node: dict, level: int, maps: dict) -> None:
            b = byte_length(stringify(node))
            kids = node.get("children") or []
            o = max_part_bytes > 0 and b > max_part_bytes
            can = rechunk and level < max_rechunk_depth and len(kids) > 0
            if (not o) or (not can):
                if o and not kids and node["id"] not in oversized:
                    oversized.append(node["id"])
                file = f"part-{sha256_of_batch([node['id']])}.json"
                emit(file, node_part(node, maps), [node["id"]])
                group_files.append(file)
                return
            counters["rechunked"] += 1
            shallow = dict(node)
            shallow["children"] = [_strip_children(c) for c in kids]
            file = f"part-{sha256_of_batch([node['id']])}.json"
            emit(file, node_part(shallow, maps), [node["id"]])
            group_files.append(file)
            for child in kids:
                split_node(child, level + 1, maps)

        for id_ in ids:
            entry = (data.get("nodes") or {}).get(id_)
            if not entry or entry.get("document") is None:
                continue
            split_node(entry["document"], depth, maps_of(entry))
        return {"sha": gsha, "ids": ids, "parts": group_files, "rechunked": True,
                "resumed": False, "localParts": local_parts, "localRechunked": counters["rechunked"], "oversized": oversized}

    results = pool([(ids, bi) for bi, ids in enumerate(batches)], concurrency, process)

    parts_meta: list[dict] = []
    groups_meta: list[dict] = []
    written = 0
    skipped = 0
    rechunked = 0
    oversized_all: list[str] = []
    seen: set[str] = set()
    for r in results:
        g = {"sha": r["sha"], "ids": r["ids"], "parts": r["parts"], "rechunked": r["rechunked"]}
        if r["resumed"]:
            g["resumed"] = True
            skipped += 1
        groups_meta.append(g)
        rechunked += r["localRechunked"]
        for m in r["localParts"]:
            if m["file"] in seen:
                continue
            seen.add(m["file"])
            if m["fresh"]:
                written += 1
            parts_meta.append({"file": m["file"], "ids": m["ids"], "sha": m["sha"], "bytes": m["bytes"], "elapsedMs": m["elapsedMs"]})
        for o in r["oversized"]:
            if o not in oversized_all:
                oversized_all.append(o)

    parts_meta.sort(key=lambda p: p["file"])
    part_files = [os.path.join(parts_dir, p["file"]) for p in parts_meta]

    manifest = {
        "fileKey": file_key,
        "size": size,
        "frontierDepth": depth,
        "batchSize": batch_size,
        "concurrency": concurrency,
        "idCount": len(frontier_ids),
        "batchCount": len(batches),
        "downloaded": written,
        "resumedSkipped": skipped,
        "rechunked": rechunked,
        "oversizedLeaves": oversized_all,
        "frontierIds": list(frontier_ids),
        "parts": parts_meta,
        "groups": groups_meta,
    }
    with open(os.path.join(work_dir, "manifest.json"), "w", encoding="utf-8") as f:
        json.dump(manifest, f, indent=2)
    return {"manifest": manifest, "partFiles": part_files}
